### PR TITLE
Orca spark estimator

### PR DESCRIPTION
**Pull request by cyita**

### CODE REVIEW

**Comment by cyita** [comment]
Jenkins passed http://10.239.47.210:18888/job/ZOO-PR-Validation/4589/

**Comment by jason-dai** [comment]
Instead of have user writing something like:
```
from zoo.orca.learn.spark_estimator import Estimator
estimator = Estimator.from_torch(...)
```

We can either keep the original user interface (using the new base class under the hood), or support something like
```
from zoo.orca.learn import Estimator
estimator = Estimator.from_tf(..., backend="bigdl")
```

File: pyzoo/zoo/orca/learn/base_estimator.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# limitations under the License.
+#
+
+from abc import ABC, abstractmethod
+
+
+class BaseEstimator(ABC):
+    @abstractmethod
+    def fit(self, data, epochs, **kwargs):
+        pass
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/base_estimator.py:23]: why pass instead of `NotImplementedError`?

File: pyzoo/zoo/orca/learn/spark_estimator.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    def predict(self, data, **kwargs):
+        pass
+
+    @abstractmethod
+    def evaluate(self, data, **kwargs):
+        pass
+
+    @abstractmethod
+    def get_model(self):
+        pass
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/spark_estimator.py:36]: why pass instead of `NotImplementedError`?

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 3]
```diff
@@ -107,7 +60,7 @@ def from_bigdl(*, model, loss=None, optimizer=None, feature_preprocessing=None,
                                      label_preprocessing=label_preprocessing, model_dir=model_dir)
 
 
-class BigDLEstimatorWrapper(Estimator):
+class BigDLEstimatorWrapper(OrcaSparkEstimator):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:63]: `set_tensorboard` already defined in `OrcaSparkEstimator`?

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 4]
```diff
@@ -558,8 +425,60 @@ def evaluate(self, data, batch_size=32,
...
+    def get_model(self):
+        pass
+
+    def save(self, model_path):
+        self.save_tf_checkpoint(model_path)
+
+    def load(self, checkpoint, **kwargs):
+        self.load_latest_orca_checkpoint(checkpoint)
+
+    def set_tensorboard(self, log_dir, app_name):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:437]: `set_tensorboard` is already defined in `SparkEstimator`?

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 5]
```diff
@@ -558,8 +425,60 @@ def evaluate(self, data, batch_size=32,
     def save_tf_checkpoint(self, path):
         save_tf_checkpoint(self.sess, path)
 
+    def get_model(self):
+        pass
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:429]: `NotImplementedError`?

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 6]
```diff
@@ -734,6 +661,59 @@ def evaluate(self, data, batch_size=4,
     def save_keras_model(self, path, overwrite=True):
         self.model.save_model(path, overwrite=overwrite)
 
+    def get_model(self):
+        pass
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:665]: NotImplementedError?

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 7]
```diff
@@ -558,8 +425,56 @@ def evaluate(self, data, batch_size=32,
...
+        self.checkpoint_version = version
+
+    def load_latest_orca_checkpoint(self, path):
+        ckpt_path, _, version = find_latest_checkpoint(path, model_type="tf")
+        if ckpt_path is None:
+            raise Exception("Cannot find checkpoint")
+        self.load_orca_checkpoint(ckpt_path, version)
+
+
+class TFKerasWrapper(SparkEstimator):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:477]: It has a lot of duplication (such as `get_train_summary`, `get_validation_summary`, etc.) with `TFOptimizerWrapper`. Maybe we should have a common base class these two classes?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 8]
```diff
@@ -296,8 +305,14 @@ def load(self, checkpoint, loss=None):
                              "and checkpoint type.")
         self.estimator = SparkEstimator(self.model, optimizer, self.model_dir)
 
-    def shutdown(self, force=False):
-        pass
+    def load_latest_orca_checkpoint(self, path):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:308]: Is this correct?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 9]
```diff
@@ -252,7 +256,7 @@ def fit(self, data, epochs=1, batch_size=32, validation_data=None, validation_me
         return self
 
     def predict(self, data, **kwargs):
-        pass
+        raise NotImplementedError
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:259]: Need to add predict later

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 10]
```diff
@@ -208,6 +207,8 @@ def __init__(self, model, loss, optimizer, model_dir=None, bigdl_type="float"):
             optimizer = SGD()
         elif isinstance(optimizer, TorchOptimizer):
             optimizer = TorchOptim.from_pytorch(optimizer)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:209]: Need to make sure `optimizer` is either `TorchOptimizer` or `or.learn.optimizers.Optimizer`

File: pyzoo/zoo/orca/learn/base_estimator.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+# limitations under the License.
+#
+
+from abc import ABC, abstractmethod
+
+
+class BaseEstimator(ABC):
+    @abstractmethod
+    def fit(self, data, epochs, **kwargs):
+        raise NotImplementedError
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/base_estimator.py:23]: `pass` will do as the subclass will need to implement all of its `abstractmethod` :-)

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 12]
```diff
@@ -207,6 +207,8 @@ def __init__(self, model, loss, optimizer, model_dir=None, bigdl_type="float"):
...
+        elif isinstance(optimizer, OptimMethod):
+            optimizer = optimizer
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:211]: 1) I meant `orca.learn.optimizers.Optimizer`
2) elif?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 13]
```diff
@@ -209,6 +210,11 @@ def __init__(self, model, loss, optimizer, model_dir=None, bigdl_type="float"):
             optimizer = TorchOptim.from_pytorch(optimizer)
         elif isinstance(optimizer, OptimMethod):
             optimizer = optimizer
+        elif isinstance(optimizer, OrcaOptimizer):
+            optimizer = optimizer.get_optimizer()
+        else:
+            raise ValueError("Only PyTorch optimizer, bigdl optimizer and orca optimizer are "
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:216]: Maybe no need to support bigdl optimizer here?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py (hunk 9 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py:263]: ok.

**Comment by cyita** [comment]
Jenkins passed http://10.239.47.210:18888/job/ZOO-PR-Validation/4783/

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 14]
```diff
@@ -204,10 +204,16 @@ def __init__(self, model, loss, optimizer, model_dir=None, bigdl_type="float"):
...
+            from zoo.orca.learn.optimizers.schedule import Default
+            optimizer = SGD(learningrate_schedule=Default())
+        if isinstance(optimizer, TorchOptimizer):
             optimizer = TorchOptim.from_pytorch(optimizer)
+        elif isinstance(optimizer, OrcaOptimizer):
+            optimizer = optimizer.get_optimizer()
+        else:
+            raise ValueError("Only PyTorch optimizer and orca optimizer are supported")
+        self.log_dir = None
+        self.app_name = None
```
Comment (hkvision) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:216]: Is there anywhere to set these two inputs? Seems they are always None and set_tensorboard will never be triggered... @cyita

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 15]
```diff
@@ -223,6 +229,9 @@ def fit(self, data, epochs=1, batch_size=32, validation_data=None, validation_me
         validation_methods = Metrics.convert_metrics_list(validation_methods)
         checkpoint_trigger = Trigger.convert_trigger(checkpoint_trigger)
 
+        if self.log_dir is not None and self.app_name is not None:
+            self.estimator.set_tensorboard(self.log_dir, self.app_name)
```
Comment (hkvision) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:233]: See the comment above

Comment (cyita) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py:233]: Please see `set_tensorboard` in spark_estimator.py.

**Pull request merged**